### PR TITLE
bump sqllogictest to 0.9.0

### PR DESCRIPTION
**Pull request by xxchan**
- bump sqllogictest & fix breaking changes: https://github.com/risinglightdb/sqllogictest-rs/blob/HEAD/CHANGELOG.md
- workaround `NULL` normalization, close https://github.com/apache/arrow-datafusion/issues/4500

re https://github.com/apache/arrow-datafusion/issues/4460

cc @alamb @xudong963

### CODE REVIEW

File: datafusion/core/tests/sqllogictests/src/main.rs [hunk 1]
```diff
@@ -172,19 +173,42 @@ async fn context_for_test_file(file_name: &str) -> SessionContext {
...
+                .map(|s| {
+                    if s.is_empty() {
+                        "NULL".to_string()
+                    } else {
+                        s.to_string()
+                    }
+                })
+                .collect()
+        })
+        .collect();
```
Comment (xxchan) [inline datafusion/core/tests/sqllogictests/src/main.rs:207]: Using tab-separated CSV to get a row as `Vec<String>` is a little bit hacky, but it works... 😄

BTW here's some information about separator in test files https://github.com/risinglightdb/sqllogictest-rs/issues/109

File: datafusion/core/tests/sqllogictests/src/main.rs [hunk 2]
```diff
@@ -172,19 +173,42 @@ async fn context_for_test_file(file_name: &str) -> SessionContext {
     }
 }
 
-fn format_batches(batches: Vec<RecordBatch>) -> Result<String> {
+fn convert_batches(batches: Vec<RecordBatch>) -> Result<DBOutput> {
     let mut bytes = vec![];
+    if batches.is_empty() {
+        return Ok(DBOutput::StatementComplete(0));
+    }
+    // TODO: use the actual types
```
Comment (alamb) [inline datafusion/core/tests/sqllogictests/src/main.rs:181]: 👍  thanks -- we can fill this in

File: datafusion/core/tests/sqllogictests/src/main.rs (hunk 1 above)
Comment (alamb) [reply datafusion/core/tests/sqllogictests/src/main.rs:207]: Thank you @xxchan -- this is great! The new interface makes a lot of sense. Thank you for the PR!

Given what I see here, I think there may be some better ways to format the output into `Vec<Vec<String>>` than  re-parsing CSV, by using arrow directly (for example how it is done in https://docs.rs/arrow/28.0.0/arrow/util/pretty/index.html)

I'll give it a think and see if I can find some way to update it 

cc @mvanschellebeeck

File: datafusion/core/tests/sqllogictests/test_files/misc.slt [hunk 3]
```diff
@@ -19,8 +19,8 @@
...
 # Show string normalization working
 # '' --> (empty)
-# '' --> NULL   (TODO: needs https://github.com/apache/arrow-rs/issues/3268)
+# '' --> NULL
 query II
 select 'foo', '', NULL
 ----
-foo (empty)
\ No newline at end of file
+foo (empty) NULL
```
Comment (xudong963) [inline datafusion/core/tests/sqllogictests/test_files/misc.slt:26]: @xxchan You are so quickly! 

I noticed your new release and prepared to remove the issue about `NULL`. Then I found you done!

Thank you

File: datafusion/core/tests/sqllogictests/test_files/ddl.slt [hunk 4]
```diff
@@ -45,9 +45,9 @@ query II rowsort
 select * from users;
 ----
 1 2
+11 20
```
Comment (xudong963) [inline datafusion/core/tests/sqllogictests/test_files/ddl.slt:48]: Why this change?

**Review by xudong963 — APPROVED** [review]
Thanks @xxchan

File: datafusion/core/tests/sqllogictests/test_files/ddl.slt (hunk 4 above)
Comment (xxchan) [reply datafusion/core/tests/sqllogictests/test_files/ddl.slt:48]: For rowsort, runner will only sort actual results now, which means the result in the test cases should be sorted.

**Comment by alamb** [comment]
FWIW if no one beats me to it I plan to try and polish the conversion to `Vec<Vec<String>>` this weekend, but I would be happy to have this one merged in first. I suspect it has been accumulating both actual conflicts (as github says) as well as logical conflicts with newly added SLT files.


I poked around in the sqllogictest runner, and it looks like you already have some code that will run / update the test files with new results that looks super awesome (and relevant to updating results)

https://github.com/risinglightdb/sqllogictest-rs/blob/d753e4c77fbf00cdaab69477df2e9c3dba3f5fcc/sqllogictest-bin/src/lib.rs#L505

**Comment by xudong963** [comment]
> I poked around in the sqllogictest runner, and it looks like you already have some code that will run / update the test files with new results that looks super awesome (and relevant to updating results)
> 
> https://github.com/risinglightdb/sqllogictest-rs/blob/d753e4c77fbf00cdaab69477df2e9c3dba3f5fcc/sqllogictest-bin/src/lib.rs#L505

Yes, it's a cool feature.



> but I would be happy to have this one merged in first

Sounds great!

**Comment by xxchan** [comment]
rebased & resolved conflicts.

**Comment by xudong963** [comment]
https://github.com/apache/arrow-datafusion/pull/4565 @xxchan 

Need to rebase master

**Comment by xxchan** [comment]
Hope this is the last fix...

**Comment by xudong963** [comment]
Let's merge!

File: datafusion/core/tests/sqllogictests/src/main.rs (hunk 1 above)
Comment (alamb) [reply datafusion/core/tests/sqllogictests/src/main.rs]: Here is a PR that avoids reparsing CSV: https://github.com/apache/arrow-datafusion/pull/4578
